src/NHibernate/Transaction/AdoNetWithSystemTransactionFactory.cs

```diff
@@ -0,0 +1,528 @@
+		/// <returns>The created transaction context.</returns>
+		protected virtual ITransactionContext CreateAndEnlistMainContext(
+			ISessionImplementor session,
+			System.Transactions.Transaction transaction)
+		{
+			var transactionContext = new SystemTransactionContext(
+				session, transaction, UseConnectionOnSystemTransactionEvents);
+			transactionContext.AmbientTransaction.EnlistVolatile(
+				transactionContext,
+				UseConnectionOnSystemTransactionEvents
+					? EnlistmentOptions.EnlistDuringPrepareRequired
+					: EnlistmentOptions.None);
+			return transactionContext;
+		}
+
+		private void EnlistDependentSession(ISessionImplementor dependentSession, ITransactionContext mainContext)
+		{
+			dependentSession.TransactionContext = CreateDependentContext(dependentSession, mainContext);
+			dependentSession.AfterTransactionBegin(null);
+		}
+
+		/// <summary>
+		/// Create a transaction context for a dependent session.
+		/// </summary>
+		/// <param name="dependentSession">The dependent session.</param>
+		/// <param name="mainContext">The context of the session owning the <see cref="ConnectionManager"/>.</param>
+		/// <returns>A dependent context for the session.</returns>
+		protected virtual ITransactionContext CreateDependentContext(ISessionImplementor dependentSession, ITransactionContext mainContext)
+		{
+			return new DependentContext(mainContext);
+		}
+
+		/// <inheritdoc />
+		public override bool IsInActiveSystemTransaction(ISessionImplementor session)
+			=> session.TransactionContext?.IsInActiveTransaction ?? false;
+
+		/// <inheritdoc />
+		public override void ExecuteWorkInIsolation(ISessionImplementor session, IIsolatedWork work, bool transacted)
+		{
+			using (var tx = new TransactionScope(TransactionScopeOption.Suppress))
+			{
+				base.ExecuteWorkInIsolation(session, work, transacted);
+				tx.Complete();
+			}
+		}
+
+		/// <summary>
+		/// Transaction context for enlisting a session with a system <see cref="System.Transactions.Transaction"/>.
+		/// It is meant for being the concrete class enlisted in the transaction.
+		/// </summary>
+		public class SystemTransactionContext : ITransactionContext, IEnlistmentNotification
+		{
+			/// <summary>
+			/// The transaction in which this context is enlisted.
+			/// </summary>
+			protected internal virtual System.Transactions.Transaction AmbientTransaction { get; }
+			/// <inheritdoc />
+			public bool ShouldCloseSessionOnSystemTransactionCompleted { get; set; }
+			/// <inheritdoc />
+			public bool IsInActiveTransaction { get; protected set; } = true;
+			/// <inheritdoc />
+			public virtual bool CanFlushOnSystemTransactionCompleted => _useConnectionOnSystemTransactionEvents;
+
+			private readonly ISessionImplementor _sessionImplementor;
+			private readonly bool _useConnectionOnSystemTransactionEvents;
+			private volatile SemaphoreSlim _semaphore;
+			private volatile bool _locked;
```

Comment:
> Also, what test demonstrates this difference?

> And the test is ..?

I do not understand. Do you want a unit test for something involving a race condition? That can be quite hard to achieve.

Or the place where the `_locked` condition is checked? That is in `Wait` and in `Lock`. The `Wait` test on `_locked` could be removed for relying on the `Lock` test, but that would cause a transaction status retrieval at each `Wait` call. Those calls are numerous. 

In those places, the `_semaphore` may be `null` with `_locked` false, or with `_locked` already true. That makes a difference. The culprit is avoiding a race condition causing the context to get locked again by a "late" thread having call `Wait` during the context disposal and then having be slow to reach the starting tests of this method. It is very unlikely to occur, and we could probably remove this and never witness the trouble. But with that test, it just causes that race to be impossible.